### PR TITLE
Add PushLimitThroughOuterJoin rule

### LINKED ISSUE #9783: Push down LIMIT through LEFT JOIN
For a query like 
```
SELECT .. FROM
a LEFT OUTER JOIN b ON a.key = b.key
ORDER BY a.something LIMIT 10
```
 the `LIMIT` can be applied before the join, as in:
```
SELECT .. FROM
(SELECT * FROM a ORDER BY a.something LIMIT 10) LEFT OUTER JOIN b ON a.key = b.key
ORDER BY a.something LIMIT 10
```

Furthermore, if we knew that `b.key` is unique, we could even eliminate the LIMIT above the JOIN, leaving only LIMIT below, on the left side.

### PR DESCRIPTION
**Pull request by kasiafi**
Fixes #9783

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Patterns.java [hunk 1]
```diff
@@ -187,6 +187,14 @@ private Patterns() {}
         }
     }
 
+    public static class Join
+    {
+        public static Property<JoinNode, JoinNode.Type> type()
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Patterns.java]: s/type/joinType

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushLimitThroughOuterJoin.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+import static com.facebook.presto.sql.planner.optimizations.QueryCardinalityUtil.extractCardinality;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.FULL;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.LEFT;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.RIGHT;
+import static com.facebook.presto.sql.planner.plan.Patterns.Join.type;
+import static com.facebook.presto.sql.planner.plan.Patterns.join;
+import static com.facebook.presto.sql.planner.plan.Patterns.limit;
+import static com.facebook.presto.sql.planner.plan.Patterns.source;
+
+public class PushLimitThroughOuterJoin
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushLimitThroughOuterJoin.java]: please add tests

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Patterns.java (hunk 1 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Patterns.java]: I think the naming is actually correct. This matches the property name -- `com.facebook.presto.sql.planner.plan.JoinNode#getType`
See also `com.facebook.presto.sql.planner.plan.Patterns.TopN#step`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushLimitThroughOuterJoin.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    @Override
+    public Result apply(LimitNode parent, Captures captures, Context context)
+    {
+        JoinNode joinNode = captures.get(CHILD);
+        PlanNode leftSource = joinNode.getLeft();
+        PlanNode rightSource = joinNode.getRight();
+
+        boolean noChange = true;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushLimitThroughOuterJoin.java]: It's easier to reason about positives. I'd structure this as:

```java
boolean changed = false;

if (...) {
   left = ...;
   changed = true;
}

if (...) {
    right = ...;
    changed = true;
}

if (changed) {
    return ...;
}

return Result.empty();
```

Alternatively, you could forgo the `changed` flag and just do:

```java
if (join.getLeft() != left || join.getRight() != right) {
    return ...;
}
```

**Comment by kasiafi** [comment]
Applied and added test.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushLimitThroughOuterJoin.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            left = new LimitNode(context.getIdAllocator().getNextId(), left, parent.getCount(), false);
+            changed = true;
+        }
+
+        if (isRightOuter(joinNode.getType()) && !isLimited(right, context.getLookup(), parent.getCount())) {
+            right = new LimitNode(context.getIdAllocator().getNextId(), right, parent.getCount(), false);
+            changed = true;
+        }
+
+        if (changed) {
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushLimitThroughOuterJoin.java]: As Martin suggested before I would inline this variable and used `left != joinNode.getLeft() || right != joinNode.getRight()`

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPushLimitThroughOuterJoin.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+                .matches(
+                        limit(1,
+                                join(
+                                        LEFT,
+                                        ImmutableList.of(equiJoinClause("leftKey", "rightKey")),
+                                        limit(1, values("leftKey")),
+                                        values("rightKey"))));
+    }
+
+    @Test
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestPushLimitThroughOuterJoin.java]: can you please add a test where rule does not need to be applied (where children produces less rows than the limit you are trying to push down.

**Comment by kasiafi** [comment]
Applied, added another test.

**Comment by kasiafi** [comment]
@kokosing @martint thanks for the review. Are there any other changes required?

**Comment by kokosing** [comment]
It looks good to me. 

Let's wait a bit for @martint, so he had a chance to see how his comments got addressed. After that I think we can merge it.

**Comment by highker** [comment]
Will merge this after release.

**Comment by highker** [comment]
Merged

**Pull request closed without merging**